### PR TITLE
NXP: Add support for Kinetis KV5x SoC series and TWR-KV58F220M development board

### PR DESCRIPTION
**Pull request by fbrozovic**
This patch series adds support for the NXP Kinetis KV5x SoC series and the TWR-KV58F220M development board.

As mentioned in the documentation, KV5x support is non-existent in pyOCD, and the OpenSDA J-Link firmware seems to be broken for KV5x targets at the moment, so an external debug probe needs to be used in order to debug the board.

### CODE REVIEW

**Comment by fbrozovic** [comment]
Not sure what's going on with the two failing tests there, they seem to be entirely unrelated to this PR (and for an entirely different board...)

File: boards/arm/twr_kv58f220m/doc/index.rst
Comment (dbkinder) [inline boards/arm/twr_kv58f220m/doc/index.rst]: Shouldn't these two references also have a ``CONFIG_`` prefix?

File: boards/arm/twr_kv58f220m/doc/index.rst
Comment (dbkinder) [inline boards/arm/twr_kv58f220m/doc/index.rst]: Can you reference the firmware version number here so the reader might notice when a newer (potentially working) version of the firmware becomes available.

File: boards/arm/twr_kv58f220m/doc/index.rst
Comment (dbkinder) [inline boards/arm/twr_kv58f220m/doc/index.rst]: Is there a bug tracking link you to add here (assuming it's a known and reported issue)

File: boards/arm/twr_kv58f220m/doc/index.rst
Comment (dbkinder) [inline boards/arm/twr_kv58f220m/doc/index.rst]:
```suggestion
An external J-Link debug probe connected to the JTAG header J13 is used
```

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/twr_kv58f220m/doc/index.rst
Comment (fbrozovic) [reply boards/arm/twr_kv58f220m/doc/index.rst]: ack

File: boards/arm/twr_kv58f220m/doc/index.rst
Comment (fbrozovic) [reply boards/arm/twr_kv58f220m/doc/index.rst]: ack

File: boards/arm/twr_kv58f220m/doc/index.rst
Comment (fbrozovic) [reply boards/arm/twr_kv58f220m/doc/index.rst]: Haven't reported it to Segger yet, so unfortunately not

File: boards/arm/twr_kv58f220m/doc/index.rst
Comment (fbrozovic) [reply boards/arm/twr_kv58f220m/doc/index.rst]: ack

**Comment by fbrozovic** [comment]
@dbkinder Comments addressed.

File: soc/arm/nxp_kinetis/kv5x/linker.ld
Comment (MaureenHelm) [inline soc/arm/nxp_kinetis/kv5x/linker.ld]: Good catch. I think we can apply this logic to other kinetis platforms like k64f and save some flash.
cc: @d3zd3z

**Review by MaureenHelm — APPROVED** [review]
Looks good, thanks @fbrozovic !

@nashif This updates mcux. Should I wait for you to pull mcux out to a module first?

cc: @MeganHansen @hakehuang

**Comment by fbrozovic** [comment]
ping @dbkinder @nashif

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, thanks.

**Comment by hakehuang** [comment]
@fbrozovic , I meet 3 issues by kernel testing see below

- 	kernel.mem_protect.mem_protect
`
======== Target Main Serial:   ========
***** Booting Zephyr OS version 1.14.99 *****

Running test suite memory_protection_test_suite

===================================================================

starting test - test_permission_inheritance

PASS - test_permission_inheritance

===================================================================

starting test - test_mem_domain_valid_access

PASS - test_mem_domain_valid_access

===================================================================

starting test - test_mem_domain_invalid_access

FAIL - test_mem_domain_invalid_access

===================================================================

starting test - test_mem_domain_partitions_user_rw

PASS - test_mem_domain_partitions_user_rw

===================================================================

starting test - test_mem_domain_partitions_supervisor_rw

PASS - test_mem_domain_partitions_supervisor_rw

===================================================================

starting test - test_mem_domain_partitions_user_ro



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/mem_protect/src/mem_domain.c:221: user_thread_entry_ro: (Reached unreachable code)

The user thread is allowed to access a read only partition of a memory domain

FAIL - test_mem_domain_partitions_user_ro

===================================================================

starting test - test_mem_domain_add_partitions_invalid



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/mem_protect/src/mem_domain.c:340: mem_domain_for_user_tc3: (Reached unreachable code)

Fault didn't occur when we accessed a unassigned memory domain.

FAIL - test_mem_domain_add_partitions_invalid

===================================================================

starting test - test_mem_domain_add_partitions_simple

PASS - test_mem_domain_add_partitions_simple

===================================================================

starting test - test_mem_domain_remove_partitions_simple



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/mem_protect/src/mem_domain.c:457: mem_domain_for_user_tc5: (Reached unreachable code)

Fault didn't occur when we accessed a unassigned memory domain.

FAIL - test_mem_domain_remove_partitions_simple

===================================================================

starting test - test_mem_domain_remove_partitions



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/mem_protect/src/mem_domain.c:498: mem_domain_test_6_2: (Reached unreachable code)

Fault didn't occur when we accessed a unassigned memory domain.

FAIL - test_mem_domain_remove_partitions

===================================================================

starting test - test_mem_domain_remove_thread



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/mem_protect/src/mem_domain.c:551: mem_domain_for_user_tc7: (Reached unreachable code)

Fault didn't occur when we accessed a unassigned memory domain.

FAIL - test_mem_domain_remove_thread

===================================================================

starting test - test_mem_domain_destroy

PASS - test_mem_domain_destroy

===================================================================

starting test - test_kobject_access_grant

0x00000000 is not a valid k_sem

syscall z_hdlr_k_sem_take failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200012bc (unknown)

Faulting instruction address = 0x100098d9

Caught system error -- reason 0 1

PASS - test_kobject_access_grant

===================================================================

starting test - test_syscall_invalid_kobject

0x20003c7c is not a valid k_sem

syscall z_hdlr_k_sem_take failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200012bc (unknown)

Faulting instruction address = 0x100098d9

Caught system error -- reason 0 1

PASS - test_syscall_invalid_kobject

===================================================================

starting test - test_thread_without_kobject_permission

thread 0x200012bc (19) does not have permission on k_sem 0x20003c58 [00000000]

syscall z_hdlr_k_sem_give failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200012bc (unknown)

Faulting instruction address = 0x100098d9

Caught system error -- reason 0 1

PASS - test_thread_without_kobject_permission

===================================================================

starting test - test_kobject_revoke_access

thread 0x20000764 (1) does not have permission on k_sem 0x20003c58 [00000000]

syscall z_hdlr_k_sem_give failed check: access denied

***** Hardware exception *****

Current thread ID = 0x20000764 (unknown)

Faulting instruction address = 0x100098d9

Caught system error -- reason 0 1

PASS - test_kobject_revoke_access

===================================================================

starting test - test_kobject_grant_access_kobj

PASS - test_kobject_grant_access_kobj

===================================================================

starting test - test_kobject_grant_access_kobj_invalid

thread 0x200007ec (2) does not have permission on k_thread 0x20000a0c [00000000]

syscall z_hdlr_k_object_access_grant failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200007ec (unknown)

Faulting instruction address = 0x10000dcb

Caught system error -- reason 0 1

PASS - test_kobject_grant_access_kobj_invalid

===================================================================

starting test - test_kobject_release_from_user

thread 0x20000874 (3) does not have permission on k_sem 0x20003c58 [00080000]

syscall z_hdlr_k_sem_give failed check: access denied

***** Hardware exception *****

Current thread ID = 0x20000874 (unknown)

Faulting instruction address = 0x100098d9

Caught system error -- reason 0 1

PASS - test_kobject_release_from_user

===================================================================

starting test - test_kobject_access_all_grant

PASS - test_kobject_access_all_grant

===================================================================

starting test - test_thread_has_residual_permissions

thread 0x200008fc (4) does not have permission on k_sem 0x20003c58 [00080000]

syscall z_hdlr_k_sem_take failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200008fc (unknown)

Faulting instruction address = 0x10000e1d

Caught system error -- reason 0 1

PASS - test_thread_has_residual_permissions

===================================================================

starting test - test_kobject_access_grant_to_invalid_thread

thread 0x200012bc (19) does not have permission on k_thread 0x200005cc [00000000]

syscall test_kobject_access_grant_to_invalid_thread failed check: access denied

PASS - test_kobject_access_grant_to_invalid_thread

===================================================================

starting test - test_kobject_access_invalid_kobject

thread 0x200012bc (19) does not have permission on k_sem 0x200005bc [00000000]

syscall z_hdlr_k_sem_take failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200012bc (unknown)

Faulting instruction address = 0x10001569

Caught system error -- reason 0 1

PASS - test_kobject_access_invalid_kobject

===================================================================

starting test - test_access_kobject_without_init_access

thread 0x200012bc (19) does not have permission on k_sem 0x200005ac [00000000]

syscall z_hdlr_k_sem_take failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200012bc (unknown)

Faulting instruction address = 0x100015b9

Caught system error -- reason 0 1

PASS - test_access_kobject_without_init_access

===================================================================

starting test - test_access_kobject_without_init_with_access

0x2000059c used before initialization

syscall z_hdlr_k_sem_take failed check: access denied

***** Hardware exception *****

Current thread ID = 0x20000654 (unknown)

Faulting instruction address = 0x10000e6d

Caught system error -- reason 0 1

PASS - test_access_kobject_without_init_with_access

===================================================================

starting test - test_kobject_reinitialize_thread_kobj

0x200006dc k_thread in use

syscall z_hdlr_k_thread_create failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200006dc (unknown)

Faulting instruction address = 0x0

Caught system error -- reason 0 1

PASS - test_kobject_reinitialize_thread_kobj

===================================================================

starting test - test_create_new_thread_from_user

PASS - test_create_new_thread_from_user

===================================================================

starting test - test_create_new_thread_from_user_no_access_stack

thread 0x20000ba4 (11) does not have permission on _k_thread_stack_element 0x20001940 [00000000]

syscall z_hdlr_k_thread_create failed check: bad stack object

***** Hardware exception *****

Current thread ID = 0x20000ba4 (unknown)

Faulting instruction address = 0x0

Caught system error -- reason 0 1

PASS - test_create_new_thread_from_user_no_access_stack

===================================================================

starting test - test_create_new_thread_from_user_invalid_stacksize

thread 0x20000984 (7) does not have permission on _k_thread_stack_element 0x20001d80 [00000000]

syscall z_hdlr_k_thread_create failed check: bad stack object

***** Hardware exception *****

Current thread ID = 0x20000984 (unknown)

Faulting instruction address = 0x0

Caught system error -- reason 0 1

PASS - test_create_new_thread_from_user_invalid_stacksize

===================================================================

starting test - test_create_new_thread_from_user_huge_stacksize

syscall z_hdlr_k_thread_create failed check: stack size 545 is too big, max is 544

***** Hardware exception *****

Current thread ID = 0x20000a94 (unknown)

Faulting instruction address = 0x0

Caught system error -- reason 0 1

PASS - test_create_new_thread_from_user_huge_stacksize

===================================================================

starting test - test_create_new_supervisor_thread_from_user

syscall z_hdlr_k_thread_create failed check: options & K_USER

***** Hardware exception *****

Current thread ID = 0x20000cb4 (unknown)

Faulting instruction address = 0x0

Caught system error -- reason 0 1

PASS - test_create_new_supervisor_thread_from_user

===================================================================

starting test - test_create_new_essential_thread_from_user

syscall z_hdlr_k_thread_create failed check: !(options & K_ESSENTIAL)

***** Hardware exception *****

Current thread ID = 0x20000984 (unknown)

Faulting instruction address = 0x0

Caught system error -- reason 0 1

PASS - test_create_new_essential_thread_from_user

===================================================================

starting test - test_create_new_higher_prio_thread_from_user

syscall z_hdlr_k_thread_create failed check: z_is_prio_lower_or_equal(prio, _current->base.prio)

***** Hardware exception *****

Current thread ID = 0x20000a94 (unknown)

Faulting instruction address = 0x0

Caught system error -- reason 0 1

PASS - test_create_new_higher_prio_thread_from_user

===================================================================

starting test - test_create_new_invalid_prio_thread_from_user

syscall z_hdlr_k_thread_create failed check: _is_valid_prio(prio, NULL)

***** Hardware exception *****

Current thread ID = 0x20000ba4 (unknown)

Faulting instruction address = 0x0

Caught system error -- reason 0 1

PASS - test_create_new_invalid_prio_thread_from_user

===================================================================

Test suite memory_protection_test_suite failed.

===================================================================
`

- 	kernel.mem_protect.protection

`
======== Target Main Serial:   ========
***** Booting Zephyr OS version 1.14.99 *****

Running test suite protection

===================================================================

starting test - exec_data

trying to call code written to 0x20000009

returned from code at 0x20000009

Execute from target buffer succeeded!



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/protection/src/main.c:151: exec_data: (Reached unreachable code)

Execute from data did not fault

FAIL - exec_data

===================================================================

starting test - exec_stack

trying to call code written to 0x20000749

returned from code at 0x20000749

Execute from target buffer succeeded!



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/protection/src/main.c:164: exec_stack: (Reached unreachable code)

Execute from stack did not fault

FAIL - exec_stack

===================================================================

starting test - exec_heap

trying to call code written to 0x200001e5

returned from code at 0x200001e5

Execute from target buffer succeeded!



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/protection/src/main.c:179: exec_heap: (Reached unreachable code)

Execute from heap did not fault

FAIL - exec_heap

===================================================================

starting test - write_ro

trying to write to rodata at 0x10005168

***** BUS FAULT *****

  Imprecise data bus error

***** Hardware exception *****

Current thread ID = 0x2000004c (unknown)

Faulting instruction address = 0x10000560

Caught system error -- reason 0

PASS - write_ro

===================================================================

starting test - write_text

trying to write to text at 0x10000640

***** BUS FAULT *****

  Imprecise data bus error

***** Hardware exception *****

Current thread ID = 0x2000004c (unknown)

Faulting instruction address = 0x10004874

Caught system error -- reason 0

PASS - write_text

===================================================================

Test suite protection failed.

===================================================================

PROJECT EXECUTION FAILED
`

- 	kernel.mem_protect.userspace 

'
***** Booting Zephyr OS version 1.14.99 *****

Running test suite userspace

===================================================================

starting test - is_usermode

PASS - is_usermode

===================================================================

starting test - write_control

PASS - write_control

===================================================================

starting test - disable_mmu_mpu

***** BUS FAULT *****

  Precise data bus error

  BFAR Address: 0x4000d000

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x100037d6

Caught system error -- reason 0

PASS - disable_mmu_mpu

===================================================================

starting test - read_kernram

0x00000000



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:224: read_kernram: (Reached unreachable code)

Read from kernel RAM did not fault

FAIL - read_kernram

===================================================================

starting test - write_kernram



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:239: write_kernram: (Reached unreachable code)

Write to kernel RAM did not fault

FAIL - write_kernram

===================================================================

starting test - write_kernro

***** BUS FAULT *****

  Imprecise data bus error

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x1000a296

Caught system error -- reason 0

PASS - write_kernro

===================================================================

starting test - write_kerntext

***** BUS FAULT *****

  Imprecise data bus error

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x1000a91e

Caught system error -- reason 0

PASS - write_kerntext

===================================================================

starting test - read_kernel_data

0



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:298: read_kernel_data: (Reached unreachable code)

Read from data did not fault

FAIL - read_kernel_data

===================================================================

starting test - write_kernel_data



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:312: write_kernel_data: (Reached unreachable code)

Write to  data did not fault

FAIL - write_kernel_data

===================================================================

starting test - read_priv_stack

-1



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:353: read_priv_stack: (Reached unreachable code)

Read from privileged stack did not fault

FAIL - read_priv_stack

===================================================================

starting test - write_priv_stack



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:379: write_priv_stack: (Reached unreachable code)

Write to privileged stack did not fault

FAIL - write_priv_stack

===================================================================

starting test - pass_user_object

0x200002f0 is not a valid k_sem

syscall z_hdlr_k_sem_init failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x100008ff

Caught system error -- reason 0

PASS - pass_user_object

===================================================================

starting test - pass_noperms_object

thread 0x200011ec (2) does not have permission on k_sem 0x20000c94 [0000]

syscall z_hdlr_k_sem_init failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x10000949

Caught system error -- reason 0

PASS - pass_noperms_object

===================================================================

starting test - start_kernel_thread

syscall z_hdlr_k_thread_create failed check: options & K_USER

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0xaaaaaaaa

Caught system error -- reason 0

PASS - start_kernel_thread

===================================================================

starting test - read_other_stack

2863311530



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:487: read_other_stack: (Reached unreachable code)

Read from other thread stack did not fault

FAIL - read_other_stack

===================================================================

starting test - write_other_stack



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:520: write_other_stack: (Reached unreachable code)

Write to other thread stack did not fault

FAIL - write_other_stack

===================================================================

starting test - revoke_noperms_object

thread 0x200011ec (2) does not have permission on k_sem 0x20000c94 [0000]

syscall z_hdlr_k_object_release failed check: object 0x20000c94 access denied

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x20000140

Caught system error -- reason 0

PASS - revoke_noperms_object

===================================================================

starting test - access_after_revoke

thread 0x200011ec (2) does not have permission on k_sem 0x200040c8 [0008]

syscall z_hdlr_k_sem_take failed check: access denied

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x10001097

Caught system error -- reason 0

PASS - access_after_revoke

===================================================================

starting test - user_mode_enter

PASS - user_mode_enter

===================================================================

starting test - write_kobject_user_pipe

syscall z_hdlr_k_pipe_get failed check: Memory region 0x200040e8 (size 1) write access denied

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x200002bc

Caught system error -- reason 0

PASS - write_kobject_user_pipe

===================================================================

starting test - read_kobject_user_pipe

syscall z_hdlr_k_pipe_put failed check: Memory region 0x200040e8 (size 1) read access denied

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x200002bc

Caught system error -- reason 0

PASS - read_kobject_user_pipe

===================================================================

starting test - access_other_memdomain



    Assertion failed at ZEPHYR_BASE/tests/kernel/mem_protect/userspace/src/main.c:646: shared_mem_thread: (Reached unreachable code)

Thread accessed global in other memory domain



FAIL - access_other_memdomain

===================================================================

starting test - domain_add_thread_drop_to_user

PASS - domain_add_thread_drop_to_user

===================================================================

starting test - domain_add_part_drop_to_user

PASS - domain_add_part_drop_to_user

===================================================================

starting test - domain_remove_part_drop_to_user

FAIL - domain_remove_part_drop_to_user

===================================================================

starting test - domain_remove_thread_drop_to_user

FAIL - domain_remove_thread_drop_to_user

===================================================================

starting test - domain_add_thread_context_switch

PASS - domain_add_thread_context_switch

===================================================================

starting test - domain_add_part_context_switch

PASS - domain_add_part_context_switch

===================================================================

starting test - domain_remove_part_context_switch

FAIL - domain_remove_part_context_switch

===================================================================

starting test - domain_remove_thread_context_switch

FAIL - domain_remove_thread_context_switch

===================================================================

starting test - test_stack_buffer

Reserved space: 32

Provided stack size: 1024

Stack object 0x20001c80[1056]

 - Testing supervisor mode

   - Thread reports buffer 0x20001ca0 size 1024

   - check read/write to stack buffer

 - Testing user mode (direct launch)

   - Thread reports buffer 0x20001ca0 size 1024

   - check read/write to stack buffer

 - Testing user mode (drop)

   - Thread reports buffer 0x20001ca0 size 1024

   - check read/write to stack buffer

Stack object 0x200020a0[1056]

 - Testing supervisor mode

   - Thread reports buffer 0x200020c0 size 1024

   - check read/write to stack buffer

 - Testing user mode (direct launch)

   - Thread reports buffer 0x200020c0 size 1024

   - check read/write to stack buffer

 - Testing user mode (drop)

   - Thread reports buffer 0x200020c0 size 1024

   - check read/write to stack buffer

Stack object 0x200024c0[1056]

 - Testing supervisor mode

   - Thread reports buffer 0x200024e0 size 1024

   - check read/write to stack buffer

 - Testing user mode (direct launch)

   - Thread reports buffer 0x200024e0 size 1024

   - check read/write to stack buffer

 - Testing user mode (drop)

   - Thread reports buffer 0x200024e0 size 1024

   - check read/write to stack buffer

Stack object 0x200028e0[1056]

 - Testing supervisor mode

   - Thread reports buffer 0x20002900 size 1024

   - check read/write to stack buffer

 - Testing user mode (direct launch)

   - Thread reports buffer 0x20002900 size 1024

   - check read/write to stack buffer

 - Testing user mode (drop)

   - Thread reports buffer 0x20002900 size 1024

   - check read/write to stack buffer

Stack object 0x20000380[1056]

 - Testing supervisor mode

   - Thread reports buffer 0x200003a0 size 1024

   - check read/write to stack buffer

 - Testing user mode (direct launch)

   - Thread reports buffer 0x200003a0 size 1024

   - check read/write to stack buffer

 - Testing user mode (drop)

   - Thread reports buffer 0x200003a0 size 1024

   - check read/write to stack buffer

PASS - test_stack_buffer

===================================================================

starting test - test_unimplemented_syscall

Unimplemented system call

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x10003c4f

Caught system error -- reason 0

PASS - test_unimplemented_syscall

===================================================================

starting test - test_bad_syscall

Bad system call id 2147483647 invoked

***** Hardware exception *****

Current thread ID = 0x200011ec (unknown)

Faulting instruction address = 0x10003c2d

Caught system error -- reason 0

PASS - test_bad_syscall

===================================================================

starting test - test_object_recycle

PASS - test_object_recycle

===================================================================

Test suite userspace failed.

===================================================================

PROJECT EXECUTION FAILED

'

**Comment by fbrozovic** [comment]
@hakehuang I think this is due to the fact that the MPU cannot protect transactions between the core and TCM, since the MPU is connected to the crossbar switch.

Currently, DTCM is used as SRAM because the smaller KV5xF512xxx24 only has DTCM and ITCM (a 64k OCRAM is present only on KV5xF1M0xxx24, which could be protected by the MPU).

Obviously, the current state is not ideal, but I don't know what the best solution would be here - any thoughts?

**Comment by hakehuang** [comment]
@fbrozovic as KV58 has two type of MPU one is original from ARM7 core, and two SMPU, so I think for TCM usage we shall configure core MPU not SMPU, @MaureenHelm, correct me if I am wrong.

**Comment by fbrozovic** [comment]
@hakehuang Ah, I see now, somehow I completely missed the core MPU in the block diagram. I will have a look at configuring it instead of the SMPU.

**Comment by fbrozovic** [comment]
@hakehuang Switched from SMPU to ARM MPU, the memory protection tests pass now.

**Comment by hakehuang** [comment]
test pass at my side. Thanks

File: dts/arm/nxp/nxp_kv5xf1m0vlx24.dtsi
Comment (galak) [inline dts/arm/nxp/nxp_kv5xf1m0vlx24.dtsi]: remove `device_type`

File: dts/arm/nxp/nxp_kv5xf512vlx24.dtsi
Comment (galak) [inline dts/arm/nxp/nxp_kv5xf512vlx24.dtsi]: remove `device_type`.

**Review by galak — CHANGES_REQUESTED** [review]
some minor nits.

**Comment by fbrozovic** [comment]
@galak comments addressed.

**Comment by fbrozovic** [comment]
I guess because of west.yml it would be best to wait for #18527 to be merged and then rebase onto master?

File: west.yml
Comment (MaureenHelm) [inline west.yml]: Need to update this now that zephyrproject-rtos/hal_nxp#3 has been merged

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: west.yml
Comment (fbrozovic) [reply west.yml]: Done

**Review by MaureenHelm — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@galak please revisit

File: dts/arm/nxp/nxp_kv5x.dtsi
Comment (galak) [inline dts/arm/nxp/nxp_kv5x.dtsi]: Is this the correct PWM for KV5x?

**Review by galak — CHANGES_REQUESTED** [review]
I don't think the PWM compatible is correct.

File: dts/arm/nxp/nxp_kv5x.dtsi
Comment (MaureenHelm) [reply dts/arm/nxp/nxp_kv5x.dtsi]: Yes

Comment (galak) [reply dts/arm/nxp/nxp_kv5x.dtsi]: Looks like it is, but maybe we should remove the prefix or have it start at pwm1 since there is a unique PWM block at 0x4003_3000

Comment (fbrozovic) [reply dts/arm/nxp/nxp_kv5x.dtsi]: Should I just increment all the pwm aliases by one?

Comment (galak) [reply dts/arm/nxp/nxp_kv5x.dtsi]: Yeah, I think that's probably best for now.

Comment (fbrozovic) [reply dts/arm/nxp/nxp_kv5x.dtsi]: Done. Hope everything is OK now.

Comment (fbrozovic) [reply dts/arm/nxp/nxp_kv5x.dtsi]: I see now that I should also increment the DT_FTM_x in dts_fixup.h then.
However, as far as I can see it's not (yet?) possible to use two different PWM drivers anyway, no? As soon as CONFIG_PWM_0 is defined, the build will fail because DT_FTM_0_xxx are not defined...

Comment (galak) [reply dts/arm/nxp/nxp_kv5x.dtsi]: > I see now that I should also increment the DT_FTM_x in dts_fixup.h then.
> However, as far as I can see it's not (yet?) possible to use two different PWM drivers anyway, no? As soon as CONFIG_PWM_0 is defined, the build will fail because DT_FTM_0_xxx are not defined...

Yeah, looks like we've got some assumptions about on a single type of PWM driver right now.

Comment (fbrozovic) [reply dts/arm/nxp/nxp_kv5x.dtsi]: Anyway, the FTM PWM instances are now numbered 1 through 4.

File: soc/arm/nxp_kinetis/kv5x/dts_fixup.h
Comment (galak) [inline soc/arm/nxp_kinetis/kv5x/dts_fixup.h]: This should start at DT_FTM_0 and go to 3.

File: dts/arm/nxp/nxp_kv5x.dtsi
Comment (galak) [inline dts/arm/nxp/nxp_kv5x.dtsi]: Make make the label FTM_0, FTM_1, etc..

**Review by galak — CHANGES_REQUESTED** [review]
Need to think a little on what to do with the aliases.

File: soc/arm/nxp_kinetis/kv5x/dts_fixup.h
Comment (fbrozovic) [reply soc/arm/nxp_kinetis/kv5x/dts_fixup.h]: Doesn't that create a mismatch between the FTM driver and the device tree, because with CONFIG_PWM_(x)=y, the FTM driver needs DT_FTM_(x)_BASE_ADDRESS (and the rest) to be defined? I presume the pwm alias numbers in device tree should correspond to CONFIG_PWM_{0,1,2...}

File: dts/arm/nxp/nxp_kv5x.dtsi
Comment (fbrozovic) [reply dts/arm/nxp/nxp_kv5x.dtsi]: ack

**Comment by fbrozovic** [comment]
@galak Any new thoughts on this?

**Comment by MaureenHelm** [comment]
> @galak Any new thoughts on this?

@galak ping

**Review by galak — APPROVED** [review]

**Comment by MaureenHelm** [comment]
@fbrozovic will you please rebase and fix the shippable failures?

**Comment by fbrozovic** [comment]
@MaureenHelm done.

**Pull request merged**
